File: metrics/src/main/java/com/facebook/battery/metrics/cpu/CpuMetrics.java
// Copyright 2004-present Facebook. All Rights Reserved.

package com.facebook.battery.metrics.cpu;

import android.support.annotation.Nullable;
import com.facebook.battery.metrics.api.SystemMetrics;

/**
 * Information about system and user cpu metrics
 */
public class CpuMetrics extends SystemMetrics<CpuMetrics> {

  private static final long serialVersionUID = 0;

  public double userTimeS;
  public double systemTimeS;
  public double childUserTimeS;
  public double childSystemTimeS;

  public CpuMetrics() {
  }

  @Override
  public CpuMetrics set(CpuMetrics metrics) {
    userTimeS = metrics.userTimeS;
    systemTimeS = metrics.systemTimeS;
    childUserTimeS = metrics.childUserTimeS;
    childSystemTimeS = metrics.childSystemTimeS;
    return this;
  }

  @Override
  public CpuMetrics sum(@Nullable CpuMetrics b, @Nullable CpuMetrics output) {
    if (output == null) {
      output = new CpuMetrics();
    }

    if (b == null) {
      output.set(this);
    } else {
      output.systemTimeS = systemTimeS + b.systemTimeS;
      output.userTimeS = userTimeS + b.userTimeS;
      output.childSystemTimeS = childSystemTimeS + b.childSystemTimeS;
      output.childUserTimeS = childUserTimeS + b.childUserTimeS;
    }

    return output;
  }

  @Override
  public CpuMetrics diff(@Nullable CpuMetrics b, @Nullable CpuMetrics output) {
    if (output == null) {
      output = new CpuMetrics();
    }

    if (b == null) {
      output.set(this);
    } else {
      output.systemTimeS = systemTimeS - b.systemTimeS;
      output.userTimeS = userTimeS - b.userTimeS;
      output.childSystemTimeS = childSystemTimeS - b.childSystemTimeS;
      output.childUserTimeS = childUserTimeS - b.childUserTimeS;
    }

    return output;
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    }
    if (other == null || getClass() != other.getClass()) {
      return false;
    }

    CpuMetrics that = (CpuMetrics) other;

    return Double.compare(that.systemTimeS, systemTimeS) == 0 &&
      Double.compare(that.userTimeS, userTimeS) == 0 &&
      Double.compare(that.childSystemTimeS, childSystemTimeS) == 0 &&
      Double.compare(that.childUserTimeS, childUserTimeS) == 0;
  }

  @Override
  public int hashCode() {
    int result;
    long temp;
    temp = Double.doubleToLongBits(systemTimeS);
    result = (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(userTimeS);
    result = 31 * result + (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(childSystemTimeS);
    result = 31 * result + (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(childUserTimeS);
    result = 31 * result + (int) (temp ^ (temp >>> 32));
    return result;
  }

  @Override
  public String toString() {
    return "CpuMetrics{" +
        "userTimeS=" + userTimeS +
        ", systemTimeS=" + systemTimeS +
        ", childUserTimeS=" + childUserTimeS +
        ", childSystemTimeS=" + childSystemTimeS +
        '}';
  }
}

File: metrics/src/main/java/com/facebook/battery/metrics/cpu/CpuMetrics.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.battery.metrics.cpu

import com.facebook.battery.metrics.core.SystemMetrics

/** Information about system and user cpu metrics: maintains a double per type of CPU time. */
class CpuMetrics : SystemMetrics<CpuMetrics>() {

  @JvmField var userTimeS: Double = 0.0

  @JvmField var systemTimeS: Double = 0.0

  @JvmField var childUserTimeS: Double = 0.0

  @JvmField var childSystemTimeS: Double = 0.0

  override fun set(metrics: CpuMetrics): CpuMetrics {
    this.userTimeS = metrics.userTimeS
    this.systemTimeS = metrics.systemTimeS
    this.childUserTimeS = metrics.childUserTimeS
    this.childSystemTimeS = metrics.childSystemTimeS
    return this
  }

  override fun sum(b: CpuMetrics?, output: CpuMetrics?): CpuMetrics {
    var output = output
    if (output == null) {
      output = CpuMetrics()
    }

    if (b == null) {
      output.set(this)
    } else {
      output.systemTimeS = systemTimeS + b.systemTimeS
      output.userTimeS = userTimeS + b.userTimeS
      output.childSystemTimeS = childSystemTimeS + b.childSystemTimeS
      output.childUserTimeS = childUserTimeS + b.childUserTimeS
    }

    return output
  }

  override fun diff(b: CpuMetrics?, output: CpuMetrics?): CpuMetrics {
    var output = output
    if (output == null) {
      output = CpuMetrics()
    }

    if (b == null) {
      output.set(this)
    } else {
      output.systemTimeS = systemTimeS - b.systemTimeS
      output.userTimeS = userTimeS - b.userTimeS
      output.childSystemTimeS = childSystemTimeS - b.childSystemTimeS
      output.childUserTimeS = childUserTimeS - b.childUserTimeS
    }

    return output
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }

    val that = other as CpuMetrics

    return java.lang.Double.compare(
        that.systemTimeS,
        systemTimeS,
    ) == 0 &&
        java.lang.Double.compare(
            that.userTimeS,
            userTimeS,
        ) == 0 &&
        java.lang.Double.compare(
            that.childSystemTimeS,
            childSystemTimeS,
        ) == 0 &&
        java.lang.Double.compare(that.childUserTimeS, childUserTimeS) == 0
  }

  override fun hashCode(): Int {
    var result: Int
    var temp = java.lang.Double.doubleToLongBits(systemTimeS)
    result = (temp xor (temp ushr 32)).toInt()
    temp = java.lang.Double.doubleToLongBits(userTimeS)
    result = 31 * result + (temp xor (temp ushr 32)).toInt()
    temp = java.lang.Double.doubleToLongBits(childSystemTimeS)
    result = 31 * result + (temp xor (temp ushr 32)).toInt()
    temp = java.lang.Double.doubleToLongBits(childUserTimeS)
    result = 31 * result + (temp xor (temp ushr 32)).toInt()
    return result
  }

  override fun toString(): String =
      "CpuMetrics{userTimeS=${userTimeS}, systemTimeS=${systemTimeS}, childUserTimeS=${childUserTimeS}, childSystemTimeS=${childSystemTimeS}${'}'}"
}
